core/tee/tee_svc.c

```diff
@@ -714,11 +718,18 @@ static TEE_Result tee_svc_update_out_param(
 				void *dst = (void *)(uintptr_t)vals[n * 2];
 				TEE_Result res;
 
-				res = tee_svc_copy_to_user(dst, src,
-						 param->u[n].mem.size);
-				if (res != TEE_SUCCESS)
-					return res;
-
+				/*
+				 * TA is allowed to return a size larger than
+				 * the original size. However, in such cases no
+				 * data should be synchronized as per TEE Client
+				 * API spec.
+				 */
+				if (param->u[n].mem.size <= tmp_buf_size[n]) {
+					res = tee_svc_copy_to_user(dst, src,
+							param->u[n].mem.size);
+					if (res != TEE_SUCCESS)
+						return res;
+				}
```

Comment:
if `(param->u[n].mem.size > tmp_buf_size[n])`, should this produce an `TEE_ERROR_SHORT_BUFFER`?

Reply:
I'm not sure it would be the right place. I'd say the called function is responsible for choosing the appropriate return code.

Reply:
Maybe the open-session sequence (line 855 in old source file), should call `tee_svc_update_out_param()` only when TA returned `SUCCESS` or `SHORT_BUFFER` only. And this function should update output values/memref buffer data only on `SUCCESS` and update memref buffer size only on `SHORT_BUFFER`.

Reply:
Well it is not my patch, and I have already adjusted it slightly for minor things, but I don't think it would be right to change it further.

The options I see here are: 1. @bsimondi reads this and proposes a different patch, 2. You propose an additional change in a separate patch, 3. We do nothing.

Reply:
Ok, looking again, I think this change here is OK but maybe `syscall_open_ta_session()` and `syscall_invoke_ta_command()` should be fixed, something like:
```Patch
@@ -xxx,x +xxx,x @@ TEE_Result syscall_open_ta_session(const TEE_UUID *dest,
 	res = tee_ta_open_session(&ret_o, &s, &utc->open_sessions, uuid,
 				  clnt_id, cancel_req_to, param);
 	tee_mmu_set_ctx(&utc->ctx);
-	if (res != TEE_SUCCESS)
+	if (res != TEE_SUCCESS || res != TEE_ERROR_SHORT_BUFFER)
 		goto function_exit;
 
-	res = tee_svc_update_out_param(param, tmp_buf_va, tmp_buf_size,
+	res2 = tee_svc_update_out_param(param, tmp_buf_va, tmp_buf_size,
 				       usr_param);
+	if (res2 != TEE_SUCCESS) {
+		/*		 *
+		 * Set origin as TEE since we failed to update output
+		 * parameters and client will not get reliable information.
+		 */
+		ret_o = TEE_ORIGIN_TEE;
+		res = res2;
+	}

 function_exit:
@@ -xxx,x +xxx,x @@ TEE_Result syscall_open_ta_session(const TEE_UUID *dest,
	res = tee_ta_invoke_command(&ret_o, called_sess, &clnt_id,
				    cancel_req_to, cmd_id, &param);
-	if (res == TEE_ERROR_TARGET_DEAD)
+	if (res != TEE_SUCCESS || res != TEE_ERROR_SHORT_BUFFER)
		goto function_exit;
 
	res2 = tee_svc_update_out_param(&param, tmp_buf_va, tmp_buf_size,
  					usr_param);
 	if (res2 != TEE_SUCCESS) {
 		/*
 		 * (...) return some error to indicate that 
 		 * parameters aren't updated as expected.
 		 */
 		ret_o = TEE_ORIGIN_TEE;
 		res = res2;
 	}
 
 function_exit
```